File: Grapevine/Cpp/code/test/test_groupUtils.cc
#include <limits.h>
#include <time.h>
#include <gtest/gtest.h>
#include <list>
#include <map>
#include <string>

#include "util.h"
#include "contextSummary.h"
#include "groupContextSummary.h"

using namespace std;

class QuickTest : public testing::Test {
 protected:
  virtual void SetUp() {
  }
  virtual void TearDown() {
  }
};

class GroupUtilsTest : public QuickTest {
 protected:
  virtual void SetUp() {
      QuickTest::SetUp();
      summary = new ContextSummary(1, db);
      summary2 = new ContextSummary(2, db2);
      summary3 = new ContextSummary(3, db3);
      groupSummary = new GroupContextSummary(101, groupDb);
      groupSummaryNull = new GroupContextSummary(100, groupDbNull);
  }
  
  virtual void TearDown() {
      QuickTest::TearDown();
      delete summary;
      delete summary2;
      delete summary3;
      delete groupSummary;
      delete groupSummaryNull;
      result.clear();
  }
  
  std::map<std::string, int> db {
        {"GroupsEnumerated",3},
        {"Group0",101},{"Group1",102},{"Group2",103},
        {"IdsAggregated",5},
        {"Id0",10}, {"Id1",20}, {"Id2",30}, {"Id3",40}, {"Id4",50}};
  std::vector<int> aggregatedIds {10,20,30,40,50};

  std::map<std::string, int> db2 {
        {"GroupsEnumerated",3},
        {"Group0",101},{"Group1",103},{"Group2",104},
        {"IdsAggregated",3},
        {"Id0",10}, {"Id1",20}, {"Id2",30}};
  std::vector<int> aggregatedIds2 {10,20,30};
  
  std::map<std::string, int> db3 {
        {"GroupsEnumerated",3},
        {"Group0",101},{"Group1",104},{"Group2",105},
        {"IdsAggregated",3},
        {"Id0",110}, {"Id1",120}, {"Id2",130}};
  std::vector<int> aggregatedIds3 {110,120,130};
  
  std::map<std::string, int> groupDb {
        {"MembersEnumerated",3},
        {"Member0",1},{"Member1",2},{"Member2",3}};
  std::vector<int> groupSummaryMembers {1,2,3};
  
  std::map<std::string, int> groupDbNull;

  ContextSummary* summary;
  ContextSummary* summary2;
  ContextSummary* summary3;
  GroupContextSummary* groupSummary;
  GroupContextSummary* groupSummaryNull;
  vector<int> result;
};

TEST_F(GroupUtilsTest, getDeclaredMemberships) {
    vector<int> result;
    vector<int> expected {101, 102, 103};
    GroupUtils::getDeclaredMemberships(*summary, result);
    //EXPECT_EQ(res, true);
    //Util::print(result);
    Util::sameTwoVectors(result, expected);
}

TEST_F(GroupUtilsTest, addDeclaredGroupMembership) {
    vector<int> result;
    vector<int> expected {100, 101, 102, 103};
    GroupUtils::addDeclaredGroupMembership(*summary, 100);
    //EXPECT_EQ(res, true);
    //Util::print(result);
    GroupUtils::getDeclaredMemberships(*summary, result);
    //EXPECT_EQ(res, true);
    Util::sameTwoVectors(result, expected);
}

TEST_F(GroupUtilsTest, declaresGroupMembership) {
    bool res = GroupUtils::declaresGroupMembership(*summary, 100);
    EXPECT_EQ(res, false);
    res = GroupUtils::declaresGroupMembership(*summary, 101);
    EXPECT_EQ(res, true);
    res = GroupUtils::declaresGroupMembership(*summary, 102);
    EXPECT_EQ(res, true);
    res = GroupUtils::declaresGroupMembership(*summary, 103);
    EXPECT_EQ(res, true);
}

TEST_F(GroupUtilsTest, addGroupMember) {
    // member is added to group
    GroupUtils::addGroupMember(*groupSummaryNull, 1200);
    //EXPECT_EQ(res, true);
    
    GroupUtils::getGroupMembers(*groupSummaryNull, result);
    std::vector<int> expected {1200};
    //EXPECT_EQ(res, true);
    EXPECT_TRUE(Util::sameTwoVectors(result, expected));
}

TEST_F(GroupUtilsTest, getGroupMembers) {
    GroupUtils::getGroupMembers(*groupSummary, result);
    std::vector<int> expected = groupSummaryMembers;
    //EXPECT_EQ(res, true);
    EXPECT_TRUE(Util::sameTwoVectors(result, expected));
}

TEST_F(GroupUtilsTest, setGroupMembersEmpty) {
    std::vector<int> emptyVector;
    // now group has nothing in it
    GroupUtils::setGroupMembers(*groupSummary, emptyVector);
    GroupUtils::getGroupMembers(*groupSummary, result);
    EXPECT_TRUE(Util::sameTwoVectors(result, emptyVector));
}

TEST_F(GroupUtilsTest, setGroupMembers) {
    std::vector<int> expected = {1,2,3,4,5,6,7,8,9,10};
    // now group has nothing in it
    GroupUtils::setGroupMembers(*groupSummary, expected);
    GroupUtils::getGroupMembers(*groupSummary, result);
    // Util::print(result);
    // Util::print(expected);
    EXPECT_TRUE(Util::sameTwoVectors(result, expected));
}

TEST_F(GroupUtilsTest, isAggregated) {
    EXPECT_TRUE(GroupUtils::isAggregated(*summary, 10));
    EXPECT_FALSE(GroupUtils::isAggregated(*summary, 1000));
}

TEST_F(GroupUtilsTest, haveNoCommonAggregation) {
    EXPECT_FALSE(GroupUtils::haveNoCommonAggregation(*summary, *summary2));
    EXPECT_TRUE(GroupUtils::haveNoCommonAggregation(*summary, *summary3));
}

TEST_F(GroupUtilsTest, getAggregatedIds) {
    GroupUtils::getAggregatedIds(*summary, result);
    EXPECT_TRUE(Util::sameTwoVectors(aggregatedIds, result));
}

TEST_F(GroupUtilsTest, addAggregatedId) {
    int ids;
    summary->get(IDS_AGGREGATED, ids);
    GroupUtils::addAggregatedId(*summary, 200);
    
    // # the ids has the total number:
    // # [1,2,3,4,5] --> 5, but the index is 0,1,2,3,4,*5
    // # So ids should be the newly assigned index
    int value;
    summary->get(ID_AGGREGATION_PREFIX + std::to_string(ids), value);
    
    int ida_plus1;
    summary->get(IDS_AGGREGATED, ida_plus1);
    EXPECT_EQ(ida_plus1, ids + 1);
    EXPECT_EQ(value, 200);
}

TEST_F(GroupUtilsTest, setAggregatedIds) {
    std::vector<int> ids {41,42,43,44,45,46,47};
    GroupUtils::setAggregatedIds(*summary, ids);
    
    int numberOfIds;
    summary->get(IDS_AGGREGATED, numberOfIds);
    EXPECT_EQ(numberOfIds, ids.size());
}

TEST_F(GroupUtilsTest, aggregateIntoGroupSummary) {
    int aggregatedSize1;
    summary->get(IDS_AGGREGATED, aggregatedSize1);
    int aggregatedSize2;
    bool res2 = groupSummary->get(IDS_AGGREGATED, aggregatedSize2);
    
    EXPECT_EQ(aggregatedIds.size(), aggregatedSize1);
    EXPECT_EQ(res2, false);
    
    GroupUtils::aggregateIntoGroupSummary(*groupSummary, *summary);
    groupSummary->get(IDS_AGGREGATED, aggregatedSize2);
    EXPECT_EQ(aggregatedIds.size(), aggregatedSize2);
}

TEST_F(GroupUtilsTest, updateGroupAggForOneSummary) {
    EXPECT_TRUE(GroupUtils::declaresGroupMembership(*summary, groupSummary->getId()));
    EXPECT_FALSE(GroupUtils::isAggregated(*groupSummary, summary->getId()));
    EXPECT_TRUE(GroupUtils::haveNoCommonAggregation(*groupSummary, *summary));
    int aggregatedSize1;
    summary->get(IDS_AGGREGATED, aggregatedSize1);
    int aggregatedSize2;
    bool res2 = groupSummary->get(IDS_AGGREGATED, aggregatedSize2);

    EXPECT_EQ(aggregatedIds.size(), aggregatedSize1);
    EXPECT_EQ(res2, false);
    
    GroupUtils::updateGroupAggForOneSummary(*groupSummary, *summary);
    groupSummary->get(IDS_AGGREGATED, aggregatedSize2);
    EXPECT_EQ(aggregatedIds.size(), aggregatedSize2);
}

TEST_F(GroupUtilsTest, updateGroupAgg) {
    std::vector<ContextSummary> summaries {*summary, *summary2, *summary3};
    GroupUtils::updateGroupAgg(*groupSummary, summaries);
    // TODO - no eq???
}

File: Grapevine/Python/test/testUtil/testGroupUtils.py
import unittest
import sys
import os.path

# src should come first as the util.groupUtils things
sys.path.append(os.path.abspath("../../src"))

#print sys.path

from util.groupUtils import *
from contextSummary import *
from groupContextSummary import *

class TestGroupUtils(unittest.TestCase):  
    def setUp(self):
        db = {"GroupsEnumerated":3,
              "Group0":101,"Group1":102,"Group2":103,
              "IdsAggregated":5,
              "Id0":10, "Id1":20, "Id2":30, "Id3":40, "Id4":50
              }
        self.summary = ContextSummary(1, db)
        self.aggregatedIds = [10,20,30,40,50]
        
        db2 = {"GroupsEnumerated":3,
              "Group0":101,"Group1":103,"Group2":104,
              "IdsAggregated":3,
              "Id0":10, "Id1":20, "Id2":30
              }
        self.summary2 = ContextSummary(2, db2)
        self.aggregatedIds2 = [10,20,30]
        
        # No common aggregation with db
        db3 = {"GroupsEnumerated":3,
              "Group0":101,"Group1":104,"Group2":105,
              "IdsAggregated":3,
              "Id0":110, "Id1":120, "Id2":130
              }
        self.summary3 = ContextSummary(3, db3)
        self.aggregatedIds3 = [110,120,130]
        
        groupDbNull = {}
        self.groupSummaryNull = GroupContextSummary(100, groupDbNull)
        
        # groupSummary that I use for test
        groupDb = {"MembersEnumerated":3,
                   "Member0":1, "Member1":2, "Member2":3}
        self.groupSummaryMembers = [1,2,3]
        self.groupSummary = GroupContextSummary(101, groupDb)
              
    def test_getDeclaredMemberships(self):
        groups = getDeclaredMemberships(self.summary)
        # http://stackoverflow.com/questions/7828867/how-to-efficiently-compare-two-unordered-lists-not-sets-in-python
        self.assertTrue(sorted(groups) == sorted([101,102,103]))
        
    def test_addDeclaredGroupMembership(self):
        self.assertFalse(declaresGroupMembership(self.summary, 100))
        groups = getDeclaredMemberships(self.summary)
        self.assertTrue(len(groups) == 3)
        
        addDeclaredGroupMembership(self.summary, 100)
        
        self.assertTrue(declaresGroupMembership(self.summary, 100))
        groups = getDeclaredMemberships(self.summary)
        self.assertTrue(len(groups) == 4)
        
    def test_declaresGroupMembership(self):
        self.assertFalse(declaresGroupMembership(self.summary, 4))
        self.assertTrue(declaresGroupMembership(self.summary, 101))
        #print self.summary
        
    def test_addGroupMember(self):
        addGroupMember(self.groupSummaryNull, 1200)
        self.assertEqual(getGroupMembers(self.groupSummaryNull), [1200])
        
    def test_getGroupMembers(self):
        members = getGroupMembers(self.groupSummaryNull)

    def test_getGroupMembers(self):
        groupDb = {}
        groupSummary = ContextSummary(100, groupDb)
        result = getGroupMembers(groupSummary)
        self.assertTrue(result == [])
        
        result = getGroupMembers(self.groupSummary)
        #print result
        self.assertTrue(sorted(result) == sorted(self.groupSummaryMembers))
        
    def test_setGroupMembers(self):
        newMembers = []
        setGroupMembers(self.groupSummary, newMembers)
        result = getGroupMembers(self.groupSummary)
        self.assertFalse(result)
        
    #def test_setGroupMembers_biggerThanBefore(self):
        newMembers = [1,2,3,4,5,6,7]
        setGroupMembers(self.groupSummary, newMembers)
        result = getGroupMembers(self.groupSummary)
        self.assertTrue(sorted(newMembers) == sorted(result))
        
        # retrieve the size of the members and compare
        size = self.groupSummary.get(MEMBERS_ENUMERATED)
        self.assertEqual(len(newMembers), size)
        
    #def test_setGroupMembers_smallerThanBefore(self):
        newMembers = [1]
        setGroupMembers(self.groupSummary, newMembers)
        result = getGroupMembers(self.groupSummary)
        self.assertTrue(sorted(newMembers) == sorted(result))
        
        size = self.groupSummary.get(MEMBERS_ENUMERATED)
        self.assertEqual(len(newMembers), size)
        
    def test_isAggregated(self):
        #isAggregated(summary, idToCheck)
        self.assertTrue(isAggregated(self.summary, 10))
        self.assertFalse(isAggregated(self.summary, 1000))
    
    def test_haveNoCommonAggregation(self):
        # summary2 has common aggregation, so false returned
        self.assertFalse(haveNoCommonAggregation(self.summary, self.summary2))
        # summary3 has no common aggregation, so true returned
        self.assertTrue(haveNoCommonAggregation(self.summary, self.summary3))
        
    def test_getAggregatedIds(self):
        result = getAggregatedIds(self.summary)
        #print self.aggregatedIds
        self.assertTrue(sorted(self.aggregatedIds) == sorted(result))
        
    def test_addAggregatedId(self):
        ids = self.summary.get(IDS_AGGREGATED)
        addAggregatedId(self.summary, 200)
        
        # the ids has the total number:
        # [1,2,3,4,5] --> 5, but the index is 0,1,2,3,4,*5
        # So ids should be the newly assigned index
        value = self.summary.get(ID_AGGREGATION_PREFIX + str(ids))
        
        self.assertEqual(self.summary.get(IDS_AGGREGATED),ids + 1)
        self.assertEqual(value, 200)
        
    def test_setAggregatedIds(self):
        ids = [41,42,43,44,45,46,47]
        setAggregatedIds(self.summary, ids)
        
        numberOfIds = self.summary.get(IDS_AGGREGATED)
        self.assertEqual(numberOfIds, len(ids))
        
    def test_aggregateIntoGroupSummary(self):
        #print self.groupSummary
        aggregatedSize1 = self.summary.get(IDS_AGGREGATED)
        aggregatedSize2 = self.groupSummary.get(IDS_AGGREGATED)
        
        self.assertEqual(len(self.aggregatedIds), aggregatedSize1)
        self.assertEqual(None, aggregatedSize2)
        
        aggregateIntoGroupSummary(self.groupSummary, self.summary)
        aggregatedSize2 = self.groupSummary.get(IDS_AGGREGATED)
        self.assertEqual(len(self.aggregatedIds), aggregatedSize2)
        #print self.groupSummary
        
    def test_updateGroupAggForOneSummary(self):
        summary = self.summary
        # print self.summary
        # 'IdsAggregated': 5, <-- 5 aggregations 
        #  'Id0': 10, 'Id4': 50, 'Id2': 30, 'Id3': 40, , 'Id1': 20
        # 'GroupsEnumerated': 3 <-- 3 groups summary is in
        # 'Group1': 102, 'Group0': 101, 'Group2': 103
        
        groupSummary = self.groupSummary
        #print groupSummary
        # {'MembersEnumerated': 3, 'Member0': 1, 'Member1': 2, 'Member2':3}
        
        # if (declaresGroupMembership(summary, gid) or summary.getId() == gid) \
        #    and not isAggregated(groupSummary, summary.getId) \
        #    and haveNoCommonAggregation(groupSummary, summary):
        
        # 1. summary should be a member 
        self.assertTrue(declaresGroupMembership(summary, groupSummary.getId()))
        # 2. not isAggregated(groupSummary, summary.getId)
        self.assertFalse(isAggregated(groupSummary, summary.getId()))
        # 3. haveNoCommonAggregation(groupSummary, summary)
        self.assertTrue(haveNoCommonAggregation(groupSummary, summary))
        
        # Then aggregateIntoGroupSummary(groupSummary, summary)
        #_updateGroupAgg(groupSummary, summary)
        # Just copied the code from test_aggregateIntoGroupSummary
        aggregatedSize1 = self.summary.get(IDS_AGGREGATED)
        aggregatedSize2 = self.groupSummary.get(IDS_AGGREGATED)
        
        self.assertEqual(len(self.aggregatedIds), aggregatedSize1)
        self.assertEqual(None, aggregatedSize2)
        updateGroupAggForOneSummary(groupSummary, summary)
        aggregatedSize2 = self.groupSummary.get(IDS_AGGREGATED)
        self.assertEqual(len(self.aggregatedIds), aggregatedSize2)
        
        # TODO
        # Check the summary.getId() == gid case
        summary = self.groupSummary
        self.assertTrue(summary.getId() == groupSummary.getId())
        
    def test_updateGroupAgg(self):
        summaries = [self.summary, self.summary2, self.summary3]
        updateGroupAgg(self.groupSummary, summaries)
        
if __name__ == "__main__":

    unittest.main(verbosity=2)
    